File: sdk-ios/Sources/MIMSAds/BannerView.swift
import UIKit

/// Banner ad view for displaying MIMS ads
public class BannerView: UIView {

    /// Slot ID for this banner
    public var slotId: String = "default"

    /// Ad size
    public var adSize: AdSize = .banner

    /// Targeting key-values
    public var targeting: [String: String] = [:]

    /// Delegate for ad events
    public weak var delegate: AdDelegate?

    private var currentAd: AdResult?
    private var imageView: UIImageView?
    private var viewabilityTimer: Timer?
    private var hasTrackedViewable = false
    private var visibleTime: TimeInterval = 0

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        return URLSession(configuration: config)
    }()

    public override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupView()
    }

    private func setupView() {
        backgroundColor = .clear
        clipsToBounds = true

        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleTap))
        addGestureRecognizer(tapGesture)
        isUserInteractionEnabled = true
    }

    /// Load an ad
    public func loadAd() {
        guard MIMSAds.shared.isInitialized else {
            delegate?.adDidFailToLoad(self, error: MIMSError.notInitialized)
            return
        }

        fetchAd { [weak self] result in
            DispatchQueue.main.async {
                guard let self = self else { return }

                switch result {
                case .success(let ad):
                    self.displayAd(ad)
                case .failure(let error):
                    self.delegate?.adDidFailToLoad(self, error: error)
                }
            }
        }
    }

    private func fetchAd(completion: @escaping (Result<AdResult, Error>) -> Void) {
        let request = AdRequest(
            slots: [AdSlot(id: slotId, width: adSize.width, height: adSize.height)],
            targeting: targeting,
            userId: MIMSAds.shared.userId,
            platform: "ios",
            country: targeting["country"] ?? ""
        )

        guard let url = URL(string: "\(MIMSAds.shared.serverUrl)/v1/ads") else {
            completion(.failure(MIMSError.invalidURL))
            return
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(MIMSAds.shared.userId, forHTTPHeaderField: "X-User-ID")

        do {
            urlRequest.httpBody = try JSONEncoder().encode(request)
        } catch {
            completion(.failure(error))
            return
        }

        session.dataTask(with: urlRequest) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let data = data else {
                completion(.failure(MIMSError.noData))
                return
            }

            do {
                let adResponse = try JSONDecoder().decode(AdResponse.self, from: data)
                if let ad = adResponse.ads.first {
                    completion(.success(ad))
                } else {
                    completion(.failure(MIMSError.noAdsAvailable))
                }
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    private func displayAd(_ ad: AdResult) {
        currentAd = ad
        hasTrackedViewable = false
        visibleTime = 0

        loadImage(from: ad.imageUrl) { [weak self] image in
            DispatchQueue.main.async {
                guard let self = self, let image = image else {
                    self?.delegate?.adDidFailToLoad(self!, error: MIMSError.imageLoadFailed)
                    return
                }

                // Create or reuse image view
                if self.imageView == nil {
                    self.imageView = UIImageView()
                    self.imageView?.contentMode = .scaleAspectFit
                    self.addSubview(self.imageView!)
                }

                self.imageView?.image = image
                self.imageView?.frame = self.bounds

                // Fire impression pixel
                self.firePixel(ad.tracking.impression)

                // Start viewability tracking
                self.startViewabilityTracking(ad)

                self.delegate?.adDidLoad(self)
            }
        }
    }

    private func loadImage(from urlString: String, completion: @escaping (UIImage?) -> Void) {
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }

        session.dataTask(with: url) { data, _, _ in
            if let data = data, let image = UIImage(data: data) {
                completion(image)
            } else {
                completion(nil)
            }
        }.resume()
    }

    private func startViewabilityTracking(_ ad: AdResult) {
        viewabilityTimer?.invalidate()
        viewabilityTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            guard let self = self, !self.hasTrackedViewable else { return }

            if self.isViewable() {
                self.visibleTime += 0.1
                if self.visibleTime >= 1.0 {
                    self.hasTrackedViewable = true
                    self.firePixel(ad.tracking.viewable)
                    self.viewabilityTimer?.invalidate()
                }
            } else {
                self.visibleTime = 0
            }
        }
    }

    private func isViewable() -> Bool {
        guard window != nil, !isHidden, alpha > 0 else { return false }

        guard let windowBounds = window?.bounds else { return false }

        let frameInWindow = convert(bounds, to: window)
        let intersection = frameInWindow.intersection(windowBounds)

        guard !intersection.isNull else { return false }

        let visibleArea = intersection.width * intersection.height
        let totalArea = bounds.width * bounds.height

        return visibleArea >= totalArea * 0.5
    }

    private func firePixel(_ urlString: String) {
        guard let url = URL(string: urlString) else { return }

        session.dataTask(with: url) { _, _, _ in
            // Ignore result
        }.resume()
    }

    @objc private func handleTap() {
        guard let ad = currentAd else { return }

        delegate?.adDidClick(self)

        if let url = URL(string: ad.tracking.click) {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
    }

    public override func layoutSubviews() {
        super.layoutSubviews()
        imageView?.frame = bounds
    }

    public override func removeFromSuperview() {
        super.removeFromSuperview()
        viewabilityTimer?.invalidate()
    }

    deinit {
        viewabilityTimer?.invalidate()
    }
}

// MARK: - Error Types

public enum MIMSError: Error, LocalizedError {
    case notInitialized
    case invalidURL
    case noData
    case noAdsAvailable
    case imageLoadFailed

    public var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "MIMS SDK not initialized. Call MIMSAds.shared.initialize() first."
        case .invalidURL:
            return "Invalid server URL"
        case .noData:
            return "No data received from server"
        case .noAdsAvailable:
            return "No ads available"
        case .imageLoadFailed:
            return "Failed to load ad image"
        }
    }
}

// MARK: - API Models

struct AdRequest: Encodable {
    let slots: [AdSlot]
    let targeting: [String: String]
    let userId: String
    let platform: String
    let country: String

    enum CodingKeys: String, CodingKey {
        case slots, targeting, platform, country
        case userId = "user_id"
    }
}

struct AdSlot: Encodable {
    let id: String
    let width: Int
    let height: Int
}

struct AdResponse: Decodable {
    let ads: [AdResult]
}

struct AdResult: Decodable {
    let slotId: String
    let impressionId: String
    let lineItemId: Int
    let creativeId: Int
    let width: Int
    let height: Int
    let imageUrl: String
    let clickUrl: String
    let tracking: Tracking

    enum CodingKeys: String, CodingKey {
        case slotId = "slot_id"
        case impressionId = "impression_id"
        case lineItemId = "line_item_id"
        case creativeId = "creative_id"
        case width, height
        case imageUrl = "image_url"
        case clickUrl = "click_url"
        case tracking
    }
}

struct Tracking: Decodable {
    let impression: String
    let viewable: String
    let click: String
}

File: sdk-android/mimsads/src/main/kotlin/com/mims/ads/BannerView.kt
package com.mims.ads

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Banner ad view for displaying MIMS ads
 */
class BannerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private var slotId: String = "default"
    private var adSize: AdSize = AdSize.BANNER
    private var targeting: Map<String, String> = emptyMap()
    private var listener: AdListener? = null

    private var currentAd: AdResult? = null
    private var imageView: ImageView? = null
    private var viewabilityJob: Job? = null
    private var hasTrackedViewable = false

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()
    private val mainHandler = Handler(Looper.getMainLooper())
    private val coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    init {
        // Set up click handling
        setOnClickListener {
            currentAd?.let { ad ->
                listener?.onAdClicked()
                openUrl(ad.tracking.click)
            }
        }
    }

    /**
     * Set the slot ID for this banner
     */
    fun setSlotId(slotId: String) {
        this.slotId = slotId
    }

    /**
     * Set the ad size
     */
    fun setAdSize(adSize: AdSize) {
        this.adSize = adSize
    }

    /**
     * Set targeting key-values
     */
    fun setTargeting(targeting: Map<String, String>) {
        this.targeting = targeting
    }

    /**
     * Set the ad listener
     */
    fun setAdListener(listener: AdListener) {
        this.listener = listener
    }

    /**
     * Load an ad
     */
    fun loadAd() {
        if (!MIMSAds.isInitialized()) {
            listener?.onAdFailedToLoad("SDK not initialized")
            return
        }

        coroutineScope.launch {
            try {
                val ad = fetchAd()
                if (ad != null) {
                    displayAd(ad)
                } else {
                    listener?.onAdFailedToLoad("No ad available")
                }
            } catch (e: Exception) {
                listener?.onAdFailedToLoad(e.message ?: "Unknown error")
            }
        }
    }

    private suspend fun fetchAd(): AdResult? = withContext(Dispatchers.IO) {
        val request = AdRequest(
            slots = listOf(
                AdSlot(
                    id = slotId,
                    width = adSize.width,
                    height = adSize.height
                )
            ),
            targeting = targeting,
            userId = MIMSAds.getUserId(),
            platform = "android",
            country = targeting["country"] ?: ""
        )

        val json = gson.toJson(request)
        val body = json.toRequestBody("application/json".toMediaType())

        val httpRequest = Request.Builder()
            .url("${MIMSAds.getServerUrl()}/v1/ads")
            .post(body)
            .header("X-User-ID", MIMSAds.getUserId())
            .build()

        val response = client.newCall(httpRequest).execute()

        if (!response.isSuccessful) {
            throw IOException("HTTP ${response.code}")
        }

        val responseBody = response.body?.string() ?: return@withContext null
        val adResponse = gson.fromJson(responseBody, AdResponse::class.java)

        adResponse.ads.firstOrNull()
    }

    private suspend fun displayAd(ad: AdResult) {
        currentAd = ad
        hasTrackedViewable = false

        // Load image
        val bitmap = loadImage(ad.imageUrl)

        withContext(Dispatchers.Main) {
            if (bitmap != null) {
                // Create or reuse ImageView
                if (imageView == null) {
                    imageView = ImageView(context).apply {
                        scaleType = ImageView.ScaleType.FIT_XY
                        layoutParams = LayoutParams(
                            LayoutParams.MATCH_PARENT,
                            LayoutParams.MATCH_PARENT
                        )
                    }
                    addView(imageView)
                }

                imageView?.setImageBitmap(bitmap)

                // Fire impression pixel
                firePixel(ad.tracking.impression)

                // Start viewability tracking
                startViewabilityTracking(ad)

                listener?.onAdLoaded()
            } else {
                listener?.onAdFailedToLoad("Failed to load ad image")
            }
        }
    }

    private suspend fun loadImage(url: String): Bitmap? = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder().url(url).build()
            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                response.body?.byteStream()?.use { inputStream ->
                    BitmapFactory.decodeStream(inputStream)
                }
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun startViewabilityTracking(ad: AdResult) {
        viewabilityJob?.cancel()
        viewabilityJob = coroutineScope.launch {
            // Check viewability every 100ms
            var visibleTime = 0L
            val requiredTime = 1000L // 1 second

            while (!hasTrackedViewable) {
                delay(100)

                if (isViewable()) {
                    visibleTime += 100
                    if (visibleTime >= requiredTime) {
                        hasTrackedViewable = true
                        firePixel(ad.tracking.viewable)
                    }
                } else {
                    visibleTime = 0
                }
            }
        }
    }

    private fun isViewable(): Boolean {
        if (!isAttachedToWindow || visibility != View.VISIBLE) {
            return false
        }

        val location = IntArray(2)
        getLocationOnScreen(location)

        val viewRect = android.graphics.Rect(
            location[0],
            location[1],
            location[0] + width,
            location[1] + height
        )

        val screenWidth = context.resources.displayMetrics.widthPixels
        val screenHeight = context.resources.displayMetrics.heightPixels

        val screenRect = android.graphics.Rect(0, 0, screenWidth, screenHeight)

        if (!android.graphics.Rect.intersects(viewRect, screenRect)) {
            return false
        }

        val intersect = android.graphics.Rect(viewRect)
        intersect.intersect(screenRect)

        val visibleArea = intersect.width() * intersect.height()
        val totalArea = width * height

        // 50% or more visible
        return visibleArea >= totalArea * 0.5
    }

    private fun firePixel(url: String) {
        coroutineScope.launch(Dispatchers.IO) {
            try {
                val request = Request.Builder().url(url).build()
                client.newCall(request).execute().close()
            } catch (e: Exception) {
                // Ignore pixel errors
            }
        }
    }

    private fun openUrl(url: String) {
        try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        } catch (e: Exception) {
            // Handle error
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        viewabilityJob?.cancel()
        coroutineScope.cancel()
    }
}

/**
 * Ad event listener
 */
interface AdListener {
    fun onAdLoaded()
    fun onAdFailedToLoad(error: String)
    fun onAdClicked()
}

// Data classes for API communication
internal data class AdRequest(
    val slots: List<AdSlot>,
    val targeting: Map<String, String>,
    @SerializedName("user_id") val userId: String,
    val platform: String,
    val country: String
)

internal data class AdSlot(
    val id: String,
    val width: Int,
    val height: Int
)

internal data class AdResponse(
    val ads: List<AdResult>
)

internal data class AdResult(
    @SerializedName("slot_id") val slotId: String,
    @SerializedName("impression_id") val impressionId: String,
    @SerializedName("line_item_id") val lineItemId: Int,
    @SerializedName("creative_id") val creativeId: Int,
    val width: Int,
    val height: Int,
    @SerializedName("image_url") val imageUrl: String,
    @SerializedName("click_url") val clickUrl: String,
    val tracking: Tracking
)

internal data class Tracking(
    val impression: String,
    val viewable: String,
    val click: String
)
